src/librbd/api/Mirror.cc

```diff
@@ -2098,6 +2269,1683 @@ void Mirror<I>::image_snapshot_create(I *ictx, uint32_t flags,
+      image_promote(ictx, group_snap_id, false, &snap_ids[i], on_finish);  // No force?
+      on_finishes[i] = on_finish;
+    }
+
+    for (int i = 0, n = image_ctxs.size(); i < n; ++i) {
+      if (!on_finishes[i]) {
+        continue;
+      }
+      r = on_finishes[i]->wait();
+      delete on_finishes[i];
+      if (r < 0) {
+        lderr(cct) << "failed promoting image: " << cpp_strerror(r)  << dendl;
+        // just report error, but don't abort the process
+      } else {
+        group_snap.snaps[i].snap_id = snap_ids[i];
+      }
+    }
+  }
+
+  group_snap.state = cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE;
+  r = cls_client::group_snap_set(&group_ioctx, group_header_oid, group_snap);
+  if (r < 0) {
+    lderr(cct) << "failed to update group snapshot metadata: "
+               << cpp_strerror(r) << dendl;
+  }
+
+  if (!quiesce_requests.empty()) {
+    util::notify_unquiesce(image_ctxs, quiesce_requests);
+  }
+
+  close_images(&image_ctxs);
+
+  return ret_code;
+}
+
+template <typename I>
+int Mirror<I>::group_resync(IoCtx& group_ioctx, const char *group_name) {
+  CephContext *cct = (CephContext *)group_ioctx.cct();
+  ldout(cct, 20) << "io_ctx=" << &group_ioctx
+		 << ", group_name=" << group_name << dendl;
+
+  std::string group_id;
+  int r = cls_client::dir_get_id(&group_ioctx, RBD_GROUP_DIRECTORY,
+				 group_name, &group_id);
+  if (r < 0) {
+    lderr(cct) << "error getting the group id: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  cls::rbd::MirrorGroup mirror_group;
+  r = cls_client::mirror_group_get(&group_ioctx, group_id, &mirror_group);
+  if (r == -ENOENT) {
+    ldout(cct, 10) << "mirroring for group " << group_name
+                   << " disabled" << dendl;
+    return -EINVAL;
+  } else if (r < 0) {
+    lderr(cct) << "failed to retrieve mirror group metadata: "
+               << cpp_strerror(r) << dendl;
+    return r;
+  } else if (mirror_group.mirror_image_mode !=
+             cls::rbd::MIRROR_IMAGE_MODE_SNAPSHOT) {
+    return -EOPNOTSUPP;
+  }
+
+  cls::rbd::MirrorSnapshotState state;
+  r = get_last_mirror_snapshot_state(group_ioctx, group_id, &state);
+  if (r == -ENOENT) {
+    state = cls::rbd::MIRROR_SNAPSHOT_STATE_NON_PRIMARY;
+    r = 0;
+  }
+  if (r < 0) {
+    return r;
+  }
+
+  if (state == cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY) {
+    lderr(cct) << "group " << group_name
+               << " is primary, cannot resync to itself" << dendl;
+    return -EINVAL;
+  }
+
+  std::vector<I *> image_ctxs;
+  r = open_group_images(group_ioctx, group_id, &image_ctxs);
+  if (r < 0) {
+    return r;
+  }
+
+  std::vector<cls::rbd::GroupImageStatus> images;
+  r = Group<I>::group_image_list_by_id(group_ioctx, group_id, &images);
+  if (r < 0) {
+    return r;
+  }
+
+  for (auto image_ctx : image_ctxs) {
+    r = image_resync(image_ctx);
+    if (r < 0 && r != -ETIMEDOUT) {
+      lderr(cct) << "previous resync failed" << cpp_strerror(r) << dendl;
+      break;
+    }
+  }
```

Comment:
Hello @trociny,

 @idryomov and @nbalacha suggested to delete the complete group and then resync it from the scratch for below reasons in short:

1. deleting and recreating the group will get a new group_id so we are sure we are not dealing with any left overs on the secondary cluster which belong to old group with the same name.
2. The image_resync is looping through only known images on secondary, it is not fetching the current image set from remote.

Have you thought about this before ?

Have not started on the implementation yet, but on the first look, deleting a group which is enabled for mirroing from secondary doesn't look clean, need to check on feasibility. The clean way would be to call group_disable followed by group_enable, but the interim states might not reflect resync'ing state straight.